### PR TITLE
Reverse proxy using base tag

### LINKED ISSUE #983: Add base URL setting for better reverse proxy support
### Is your proposal related to a problem?

My production server has been given a subdomain but not *.subdomain. I can't control that, so my reverse proxy forwards requests to various apps based on subdirectory. Tooljet doesn't have base url setting so I can't use it in prod properly.  

### Describe the solution you'd like

Add a base URL setting. For example that's how sonarr(a popular self-hosted PVR for series and tv shows) does it:
image

### PR DESCRIPTION
**Pull request by Navaneeth-pk**
Fixes #983

### CODE REVIEW

File: server/src/app.module.ts [hunk 1]
```diff
@@ -34,6 +34,7 @@ import { LibraryAppModule } from './modules/library_app/library_app.module';
 import { ThreadModule } from './modules/thread/thread.module';
 import { EventsModule } from './events/events.module';
 import { GroupPermissionsModule } from './modules/group_permissions/group_permissions.module';
+import * as fs from 'fs';
```
Comment (gondar00) [inline server/src/app.module.ts:37]: Probably can look at adding `fs/promises` instead

File: server/src/app.module.ts [hunk 2]
```diff
@@ -86,6 +87,28 @@ const imports = [
...
+  for (const fileName of filesToReplaceAssetPath) {
+    const file = join(__dirname, '../../../', 'frontend/build', fileName);
+
+    let newValue = process.env.SUB_PATH;
+
+    if (process.env.SUB_PATH === undefined) {
+      newValue = fileName === 'index.html' ? '/' : '';
+    }
+
+    fs.readFile(file, 'utf8', function (err, data) {
```
Comment (gondar00) [inline server/src/app.module.ts:101]: Probably need to use a sync function here? 

```js
const data = await fs.readFile(file, 'utf8')
...
await fs.writeFile(file, result, 'utf8')
```

File: frontend/src/Editor/RealtimeEditor.jsx [hunk 3]
```diff
@@ -14,9 +14,21 @@ const ydoc = new Y.Doc();
...
+
+  if (apiUrlStartsWithProtocol) {
+    url = `${config.apiUrl.replace(/(^\w+:|^)\/\//, '').replace('/api', '')}`;
+  } else {
+    url = `${window.location.host}${config.apiUrl.replace(/(^\w+:|^)\/\//, '').replace('/api', '')}`;
+  }
+
+  if (isSecure) return `${protocol}://${url}`;
+
+  return `${protocol}://${url}`;
```
Comment (gondar00) [inline frontend/src/Editor/RealtimeEditor.jsx:31]: Over here regardless of `isSecure` its returning the same thing

File: frontend/src/App/App.jsx [hunk 4]
```diff
@@ -1,6 +1,6 @@
 import React from 'react';
 import config from 'config';
-import { Router, Route, Redirect } from 'react-router-dom';
+import { BrowserRouter, Router, Route, Redirect } from 'react-router-dom';
```
Comment (gondar00) [inline frontend/src/App/App.jsx:3]: Don't think `Router` is required anymore

**Review by sherfin94 — APPROVED** [review]

**Review by akshaysasidrn — APPROVED** [review]

**Pull request merged**